src/os/bluestore/BlueStore.cc

```diff
@@ -8287,56 +8287,189 @@ int BlueStore::repair_omap_upgrade()
-        }
-        cnt++;
-        if (cnt >= 1024) {
-          key = it->key();
-          db->submit_transaction_sync(txn);
-          it->lower_bound(key);
-          txn = db->get_transaction();
-          cnt = 0;
-        }
+	dout(20) << __func__ << " key=" << pretty_binary_string(key) << dendl;
+	it->next();
+	ceph_assert(key.length() >= 8 + 4 + 8 + 1);
+	const char* p = key.c_str();
+	uint64_t pool;
+	unsigned hash;
+	uint64_t nid;
+	p = _key_decode_u64(p, &pool);
+	p = _key_decode_u32(p, &hash);
+	p = _key_decode_u64(p, &nid);
+	if (pool == prev_pool && hash == prev_hash && nid == prev_nid) {
+	  if (expect_next_object) {
+	    dout(10) << __func__ << " next object expected, but old continued" << dendl;
+	    problems++;
+	  }
+	} else {
+	  if (!expect_next_object) {
+	    // we should continue same object
+	    dout(10) << __func__ << " next objext unexpected" << dendl;
+	    problems++;
+	    expect_header_copy = false;
+	  }
+	  prev_pool = pool;
+	  prev_hash = hash;
+	  prev_nid = nid;
+	}
+	char c = *p;
+	p++;
+
+	if (expect_header_copy) {
+	  // copy of header should look like:
+	  // pppppppphhhhnnnnnnnn.ppppppppnnnnnnnn-   (for PERPOOL->PERPG failed upgrade)
+	  // pppppppphhhhnnnnnnnn.nnnnnnnn-           (for BULK->PERPG failed upgrade)
+
+	  // we expect copy only right after valid header
+	  expect_header_copy = false;
+	  if (key.length() == 8 + 4 + 8 + 1 + 8 + 8 + 1) {
+	    uint64_t pool_copy;
+	    uint64_t nid_copy;
+	    char c1;
+	    const char* p1 = p;
+	    p1 = _key_decode_u64(p1, &pool_copy);
+	    p1 = _key_decode_u64(p1, &nid_copy);
+	    c1 = *p1;
+	    if (c == '.' && pool_copy == pool && nid_copy == nid && c1 == '-') {
+	      // bingo - delete
+	      dout(15) << __func__ << " rmkey " << pretty_binary_string(key) << dendl;
+	      txn->rmkey(PREFIX_PERPG_OMAP, key);
+	      cnt += 1000;
+	      continue;
+	    }
+	  }
+	  if (key.length() == 8 + 4 + 8 + 1 + 8 + 1) {
+	    uint64_t nid_copy;
+	    char c1;
+	    const char* p1 = _key_decode_u64(p, &nid_copy);
+	    c1 = *p1;
+	    if (c == '.' && nid_copy == nid && c1 == '-') {
+	      // bingo - delete
+	      dout(15) << __func__ << " rmkey " << pretty_binary_string(key) << dendl;
+	      txn->rmkey(PREFIX_PERPG_OMAP, key);
+	      cnt += 1000;
+	      continue;
+	    }
+	  }
+	  dout(10) << __func__ << " expected copy of header. not found" << dendl;
+	  problems++;
+	}
+
+	if (c == '~') {
+	  // Tail. Good place to submit transaction.
+	  // We need to wait for omap tail before writing to DB
+	  // to avoid iterating over keys we just processed.
+	  if (cnt > 1000000) {
+	    if (dry_run) {
+	      dout(10) << __func__ << " dropping DB transaction" << dendl;
+	    } else {
+	      dout(10) << __func__ << " submit DB" << dendl;
+	      db->submit_transaction_sync(txn);
+	    }
+	    it->upper_bound(key);
+	    txn = db->get_transaction();
+	    cnt = 0;
+	  }
+	  // Tail key should not have extra chars.
+	  if (key.length() != 8 + 4 + 8 + 1) {
```

Comment:
IMO this should go before submitting a transaction to avoid false submissions. It should just cause continuation as any other unexpected record.

Reply:
And it apparently shouldn't set expect_next_object in this case